python_ta/checkers/possibly_undefined_checker.py

```diff
@@ -0,0 +1,119 @@
+    msgs = {'E9969': ('This variable might not be defined when this statement is executed.',
+                      'possibly-undefined',
+                      'Reported when a statement uses a variable that might not be assigned.'),
+           }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    def __init__(self, linter=None):
+        super().__init__(linter=linter)
+        self._possibly_undefined: Set[astroid.Name] = set()
+
+    @check_messages('possibly-undefined')
+    def visit_name(self, node):
+        """Adds message if there exists a path from the start block to node where
+        a variable used by node might not be defined."""
+        if node in self._possibly_undefined:
+            self.add_message('possibly-undefined', node=node)
+
+    def visit_module(self, node):
+        self._analyze(node)
+
+    def visit_functiondef(self, node):
+        self._analyze(node)
+
+    def _analyze(self, node: Union[astroid.Module, astroid.FunctionDef]) -> None:
+        """Runs the data flow algorithm on a `Module` or `Function` CFG, which in turn
+        appends `Name` nodes to `self.possibly_undefined` if it might not be defined.
+
+        Data flow algorithms retrieved from:
+        https://www.seas.harvard.edu/courses/cs252/2011sp/slides/Lec02-Dataflow.pdf#page=31
+        """
+        out_facts = {}
+        blocks = list(node.cfg.get_blocks_postorder())
+        blocks.reverse()
+
+        all_assigns = self._get_assigns(node)
+        for block in blocks:
+            out_facts[block] = {}
+            out_facts[block] = all_assigns.copy()
+
+        worklist = blocks
+        while len(worklist) != 0:
+            b = worklist.pop()
+            outs = [out_facts[p.source] for p in b.predecessors if p.source in out_facts]
+            if outs == []:
+                in_facts = set()
+            else:
+                in_facts = set.intersection(*outs)
+            temp = self._transfer(b, in_facts, all_assigns)
+            if temp != out_facts[b]:
+                out_facts[b] = temp
+                worklist.extend([succ.target for succ in b.successors])
+
+    def _transfer(self, block: CFGBlock, in_facts: Set[str], local_vars: Set[str]) -> Set[str]:
+        gen = in_facts.copy()
+        kill = set()
+        for statement in block.statements:
+            if isinstance(statement, astroid.FunctionDef):
+                continue
+            for node in statement.nodes_of_class((astroid.AssignName, astroid.DelName, astroid.Name),
+                                              astroid.FunctionDef):
+                if isinstance(node, astroid.AssignName):
+                    gen.add(node.name)
+                elif isinstance(node, astroid.DelName):
+                    kill.add(node.name)
+                else:
+                    name = node.name
+                    if not self._is_function_name(node) and name in local_vars \
```

Comment:
I did some more digging about this builtin function issue to see what pylint does. The short version is that I want you to replace `_is_function_name` with the three checks done [here](https://github.com/PyCQA/pylint/blob/master/pylint/checkers/variables.py#L1660). It's also worth your time to read through that `visit_name` function to look at the different cases pylint already handles. In particular, good to think about recursive function calls and using the class name in a type annotation.

Reply:
I wasn't able to add one [check](https://github.com/PyCQA/pylint/blob/6b3afd4f6d75debd7f286f0d3c760ed10ab1e79f/pylint/checkers/variables.py#L1662) from the three you mentioned. I wasn't able to find the `additional_builtins` attribute definition in the entire pylint project. According to the documentation, this attribute stores additional user-defined builtins and by default it an empty tuple. 

Either way, there's still an issue with user-defined function names that conflict with variable names.

There is currently a failing test: 
https://github.com/pyta-uoft/pyta/blob/759455aa030288e470b5167a0ef64a39df452e5e/tests/test_custom_checkers/test_possibly_undefined_checker.py#L162-L176

Reply:
The `additional_builtins` is a pylint configuration option (see [here](http://pylint.pycqa.org/en/latest/technical_reference/features.html#variables-checker-options); these are actually added dynamically from config files (e.g., `.pylintrc`), which is why you can't find it in the pylint project. You should add it in.

I'm not sure exactly what you're pointing out with the test there. Is the problem that the test as written is _passing_ but it should actually be failing?

Reply:
This test as written is not passing but if the implementation were correct (implying it currently isn't) it should be passing. The reason is that `y` in `y()` is a `Name(name=y)` node but there isn't a check to see if it's a function name so the algorithm thinks that a variable (that is not definitely assigned) is being used. 

The checks you have mentioned would be enough to filter out builtin function call name nodes however when it comes to user-defined functions, these kinds of cases can occur.

Reply:
I see. You should keep in mind that functions aren't that special---variables can refer to functions, and this is perfectly okay. So for example you could change the test to `y = lambda: 20` and the function type-checks properly, except that `y` might not be defined.